src/geo/leaflet/leaflet-vector-cartodb-group-layer-base.js

```diff
@@ -0,0 +1,309 @@
+    this._adjustTilePoint(tilePoint);
+
+    var tiles = [EMPTY_GIF];
+    if(this.tilejson) {
+      tiles = this.tilejson.tiles;
+    }
+
+    var index = (tilePoint.x + tilePoint.y) % tiles.length;
+
+    return L.Util.template(tiles[index], L.Util.extend({
+      z: this._getZoomForUrl(),
+      x: tilePoint.x,
+      y: tilePoint.y
+    }, this.options));
+  },
+
+  /**
+   * Change opacity of the layer
+   * @params {Integer} New opacity
+   */
+  setOpacity: function(opacity) {
+
+    if (isNaN(opacity) || opacity>1 || opacity<0) {
+      throw new Error(opacity + ' is not a valid value');
+    }
+
+    // Leaflet only accepts 0-0.99... Weird!
+    this.options.opacity = Math.min(opacity, 0.99);
+
+    if (this.options.visible) {
+      L.TileLayer.prototype.setOpacity.call(this, this.options.opacity);
+      this.fire('updated');
+    }
+  },
+
+
+  /**
+   * When Leaflet adds the layer... go!
+   * @params {map}
+   */
+  onAdd: function(map) {
+    // this.options.user = this.options.user_name;
+    // this.options.table = this.options.layer_definition.layers[0].options.layer_name;
+    var self = this;
+    this.options.map = map;
+
+    // Add cartodb logo
+    if (this.options.cartodb_logo != false)
+      CartoDBLogo.addWadus({ left:8, bottom:8 }, 0, map._container);
+    // TODO: We can probably move this to an initialize method
+    this.model.bind('change:urls', function() {
+      self.__update(function() {
+        // if while the layer was processed in the server is removed
+        // it should not be added to the map
+        var id = L.stamp(self);
+        if (!map._layers[id]) {
+          return;
+        }
+        L.CartoDBd3Layer.prototype.onAdd.apply(self, [map]);
+        L.CartoDBd3Layer.prototype.setCartoCSS.apply(self, [self.model.collection.models[1].layers.models[0].get("cartocss")]); 
+        self.fire('added');
+        self.options.added = true;
+      });
+    });
+  },
+
+
+  /**
+   * When removes the layer, destroy interactivity if exist
+   */
+  onRemove: function(map) {
+    if(this.options.added) {
+      this.options.added = false;
+      L.TileLayer.prototype.onRemove.call(this, map);
+    }
+  },
+
+  /**
+   * Update CartoDB layer
+   * generates a new url for tiles and refresh leaflet layer
+   * do not collide with leaflet _update
+   */
+  __update: function(done) {
+    var self = this;
+    this.fire('updated');
+    this.fire('loading');
+    var map = this.options.map;
+    var tilejson = self.model.get('urls');
+    if(tilejson) {
+      self.tilejson = tilejson;
+      self.setUrl(self.tilejson.tiles[0]);
+      // manage interaction
+      self._reloadInteraction();
+      self.ok && self.ok();
+      done && done();
+    } else {
+      self.error && self.error(err);
+      done && done();
+    }
+  },
+
+
+  _checkLayer: function() {
+    if (!this.options.added) {
+      throw new Error('the layer is not still added to the map');
```

Comment:
Wouldn't be better to return a Boolean? Not a fan of using exceptions for control flow.